### PR TITLE
Applicative Gen.apply

### LINKED ISSUE #397: Applicative behavior for Gen.apply
Currently the behavior of `Gen.apply` is monadic, but it should be applicative.

https://github.com/hedgehogqa/fsharp-hedgehog/blob/76787a5f2e7414913853bf981dd9855fcb0684d2/src/Hedgehog/Gen.fs#L55-L57

### PR DESCRIPTION
**Pull request by TysonMN**
Resolves #397

...at least that is the plan.  Currently this is a draft PR.  The behavior is not correct yet.  There is a failing test.

### CODE REVIEW

**Comment by TysonMN** [comment]
Part of this PR reverts the runtime optimization added in PR #239 because correct applicative behavior is more important (since it can lead to smaller shrunken values).

**Comment by TysonMN** [comment]
I force pushed a new stack of commits.  They are essentially the same, but some things should now be clearer.  The main improvement is the addition of a commit that adds a characterization test for `Gen.apply`.

**Review by dharmaturtle — APPROVED** [review]

File: tests/Hedgehog.Tests/GenTests.fs [hunk 1]
```diff
@@ -135,4 +135,52 @@ let genTests = testList "Gen tests" [
...
+            |> Gen.apply (Range.constant 0 2 |> Gen.int32)
+            |> Gen.apply (Range.constant 0 1 |> Gen.int32)
+
+        let actual =
+            seq {
+                while true do
+                    let t = gPair |> Gen.sampleTree 0 1 |> Seq.head
+                    if Tree.outcome t = (2, 1) then
+                        yield t
+            } |> Seq.head
```
[the remark this answers is not in the document]
Comment (TysonMN) [reply tests/Hedgehog.Tests/GenTests.fs:158]: I am satisfied with what I wrote for now.

I would prefer to have a function called something like `Seq.generate` that is similar to [`Seq.replicate`](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-seqmodule.html#replicate) but takes a function with signature `unit -> 'a` and returns `Seq<'a>` with infinite length.

File: tests/Hedgehog.Tests/TreeTests.fs [hunk 2]
```diff
@@ -142,4 +142,42 @@ let treeTests = testList "Tree tests" [
             Expect.isTrue (expected = Tree.renderList tree)
         }
         |> Property.check
+
+    testCase "apply behavior is applicative (not monadic)" <| fun () ->
```
[the remark this answers is not in the document]
Comment (TysonMN) [reply tests/Hedgehog.Tests/TreeTests.fs:146]: Sure.  Added [this link](https://well-typed.com/blog/2019/05/integrated-shrinking#example-generating-pairs).

**Comment by TysonMN** [comment]
I squashed the two changes based on your comments into the "correct" commits and force pushed

**Pull request merged**